### PR TITLE
[Refactor] Unification of buses

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | yes
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 1]
```diff
@@ -33,14 +33,21 @@ api_platform:
...
-                    - 'validation'
-                    - 'doctrine_transaction'
+            sylius.command_bus: &command_bus
+                default_middleware: allow_no_handlers
+            sylius.event_bus: &event_bus
+                default_middleware: allow_no_handlers
+            sylius_default.bus: *command_bus
+            sylius_event.bus: *event_bus
+            
+services:
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:43]: Why did you decide on an additional YAML services definition?

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 2]
```diff
@@ -33,14 +33,21 @@ api_platform:
...
-            sylius_default.bus:
-                middleware:
-                    - 'validation'
-                    - 'doctrine_transaction'
-            sylius_event.bus:
-                middleware:
-                    - 'validation'
-                    - 'doctrine_transaction'
+            sylius.command_bus: &command_bus
+                default_middleware: allow_no_handlers
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:37]:
```suggestion
                middleware: 
                    - 'validation'
                    - 'doctrine_transaction'
```

File: adr/2021_05_06_command_bus_unification.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+* Bad, because we need to keep track of all available buses
+
+### Unifying buses across all products
+
+* Good, because we will have only few buses to keep track of
+* Bad, because potential bc break
+
+## Decision Outcome
+
+Chosen option: Unifying buses across all products, it allows us stick to one bus naming style
```
Comment (lchrusciel) [inline adr/2021_05_06_command_bus_unification.md:24]:
```suggestion
Sylius and any other following product (like plugins) should utilize `sylius.command_bus` for command dispatching and `sylius.event_bus` for events.

Take into account, that the command bus requires to have one corresponding handler for each message dispatched in contradiction to the event bus(where there is no such requirement). In addition, the command bus will perform command validation and wrap the following handler within a transaction and flush to the database. 
```

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml (hunk 1 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:43]: that was only way to mark them as deprecated

File: adr/2021_05_06_command_bus_unification.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Unification of buses in sylius
+
+* Status: accepted
+* Date: 2021-05-06
+
+## Context and Problem Statement
+
+Sylius and any other following product (like plugins) should utilize `sylius.command_bus` for command dispatching and `sylius.event_bus` for events.
+
+Take into account, that the command bus requires to have one corresponding handler for each message dispatched in contradiction to the event bus(where there is no such requirement).
```
Comment (GSadee) [inline adr/2021_05_06_command_bus_unification.md:10]:
```suggestion
Take into account, that the command bus requires to have one corresponding handler for each message dispatched in contradiction to the event bus (where there is no such requirement). 
```

**Review by GSadee — COMMENTED** [review]
What about changes in CoreBundle: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/CoreBundle/Resources/config/app/messenger.yaml#L16:L18 ?

The `sylius_default.bus` is configured in Core as a default bus, but in ApiBundle it is marked as deprecated

**Comment by stloyd** [comment]
This still is broken when your existing application has own bus declared, see https://github.com/Sylius/Sylius/pull/12561.

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 5]
```diff
@@ -33,14 +33,17 @@ api_platform:
...
-            sylius_default.bus:
+            sylius.command_bus: &command_bus
                 middleware:
                     - 'validation'
                     - 'doctrine_transaction'
-            sylius_event.bus:
+            sylius.event_bus: &event_bus
                 middleware:
                     - 'validation'
                     - 'doctrine_transaction'
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:43]:
```suggestion
                    - 'allow_no_handler'
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 6]
```diff
@@ -33,17 +33,17 @@ api_platform:
...
     messenger:
         buses:
-            sylius_default.bus:
-                middleware:
-                    - 'validation'
-                    - 'doctrine_transaction'
-            sylius_event.bus:
-                middleware:
-                    - 'validation'
-                    - 'doctrine_transaction'
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:43]: Are these removed middlewares configured somewhere?

File: config/services.yaml [hunk 7]
```diff
@@ -2,3 +2,6 @@
 # https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration
 parameters:
     locale: en_US
+    
+services:
+    api_platform.message_bus: '@sylius_default.bus'
```
Comment (lchrusciel) [inline config/services.yaml:7]: This should be part of ApiBundle config and we should use XML notation for that 
```xml
<service id="api_platform.message_bus" alias="sylius_default.bus">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml (hunk 6 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:43]: Only event bus middlewares were meant to be removed. Event bus should be used only to inform the rest of the system about already happened stuff, so there is no need for transaction nor validation. Any new logic should be put on the command bus again if some part of the system would like to react to the event

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 8]
```diff
@@ -33,17 +33,17 @@ api_platform:
...
+            sylius.command_bus: &command_bus
+                default_middleware: allow_no_handlers
+            sylius.event_bus: &event_bus
+                default_middleware: allow_no_handlers
```
[the remark this answers is not in the document]
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:39]: yes, application crashes with middleware notation

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml [hunk 9]
```diff
@@ -28,6 +28,19 @@ imports:
 parameters:
     sylius_core.public_dir: "%kernel.project_dir%/web"
 
+framework:
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml:31]: This config should be moved to `src/Sylius/Bundle/CoreBundle/Resources/config/app/messenger.yaml`

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 10]
```diff
@@ -31,6 +31,8 @@
             <argument type="service" id="api_platform.iri_converter" />
         </service>
 
+        <service id="api_platform.message_bus" alias="sylius_default.bus" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml:34]: What is this line needed for?

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/messenger.yaml [hunk 11]
```diff
@@ -13,6 +13,14 @@ framework:
...
         buses:
-            sylius_default.bus: ~
+            sylius.command_bus: &command_bus
+                middleware:
+                    - 'validation'
+                    - 'doctrine_transaction'
+            sylius.event_bus: &event_bus
+                default_middleware: allow_no_handlers
+            sylius_default.bus: *command_bus
+            sylius_event.bus: *event_bus
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/messenger.yaml:25]: I'm not sure if these aliases work properly, did you check that? It looks like new buses are creating here

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 12]
```diff
@@ -195,15 +195,10 @@
...
-            <argument type="service" id="sylius_default.bus" />
+            <argument type="service" id="sylius.command_bus" />
             <tag name="kernel.event_listener" event="security.interactive_login" method="onInteractiveLogin" />
         </service>
 
-        <service id="sylius.listener.api_authentication_success_listener" class="Sylius\Bundle\ApiBundle\EventListener\AuthenticationSuccessListener">
-            <argument type="service" id="api_platform.iri_converter" />
-            <tag name="kernel.event_listener" event="lexik_jwt_authentication.on_authentication_success" method="onAuthenticationSuccessResponse" />
-        </service>
-
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml:206]: Why did you remove this listener?

File: composer.json [hunk 13]
```diff
@@ -164,7 +164,8 @@
         "symfony/doctrine-bridge": "4.4.16",
         "symfony/polyfill-mbstring": "^1.22.0",
         "symfony/property-info": "4.4.22 || 5.2.7",
-        "symfony/serializer": "4.4.19 || 5.2.2"
+        "symfony/serializer": "4.4.19 || 5.2.2",
+        "symfony/security-core": "^5.2.6 || ^4.4.21"
```
Comment (GSadee) [inline composer.json:168]: Should be reverted

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, @SirDomin! :1st_place_medal:
